### PR TITLE
[bitnami/contour] Adding the Contour Helm Chart.

### LINKED ISSUE #2030: {feature request} Onboard [bitnami/metallb] and [bitnami/contour] helm charts.
## Why?

Bitnami provides easy to extend and use Helm Charts.
I want to provide the same for MetalLB and Contour.
We use Bitnami Helm charts heavily in our own projects.

## Clarification of onboarding process

* What should be provided from me?
* What should be looked out for when onboarding?
* What are the requirements?
* Are their templates for the bitnami-docker-\<software\> repositories?

## metallb
A LoadBalancer for Kubernetes.

* My current chart implementation
[release/charts/metallb/README.md](https://github.com/Cellebyte/helm/blob/release/charts/metallb/README.md)
* The project.
[MetalLB](https://github.com/metallb/metallb/)

## contour
A IngressController for Kubernetes with CRDs.

* My current chart implementation
[release/charts/contour/README.md](https://github.com/Cellebyte/helm/blob/release/charts/contour/README.md)
* The project.
[Contour](https://github.com/projectcontour/contour)

### PR DESCRIPTION
**Pull request by Cellebyte**
**Description of the change**

Added the contour helm chart to the bitnami repository

**Benefits**
Support for a production grade contour deployment.

**Possible drawbacks**

**Applicable issues**

  - partially fixes #2030

**Additional information**

**Checklist** 
- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [x] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

### CODE REVIEW

File: bitnami/contour/values.yaml [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  # disable ingressroute permitInsecure field
+  disablePermitInsecure: false
+  tls:
+  #   minimum TLS version that Contour will negotiate
+  #   minimum-protocol-version: "1.1"
+  # The following config shows the defaults for the leader election.
+  ## This needs to be edited by when you deploy to a namespace other than projectcontour
+  # leaderelection:
+  #   configmap-name: leader-elect
+  #   configmap-namespace: projectcontour
```
Comment (Cellebyte) [inline bitnami/contour/values.yaml]: Do you have an advice on how to handle the namespace dependant on the deployed namespace properly.
I am currently thinking of tplValue from other charts.

File: bitnami/contour/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+icon: https://d33wubrfki0l68.cloudfront.net/b92d8706535e6f2b71f90645fe9ab7b418fb8ca8/ea52e/img/contour.svg
+sources:
+- https://github.com/projectcontour/contour
+- https://github.com/envoyproxy/envoy
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+version: 0.1.3
+maintainers:
+  - name: cellebyte
+    url: https://github.com/Cellebyte
```
Comment (carrodher) [inline bitnami/contour/Chart.yaml]:
```suggestion
apiVersion: v1
name: contour
description: Contour Ingress controller for Kubernetes
appVersion: 1.2.1
keywords:
  - ingress
  - envoy
  - contour
home: https://projectcontour.io
icon: https://d33wubrfki0l68.cloudfront.net/b92d8706535e6f2b71f90645fe9ab7b418fb8ca8/ea52e/img/contour.svg
sources:
  - https://github.com/projectcontour/contour
  - https://github.com/envoyproxy/envoy
  - https://github.com/bitnami/bitnami-docker-contour
version: 0.1.3
maintainers:
  - name: cellebyte
    url: https://github.com/Cellebyte
  - name: Bitnami
    email: containers@bitnami.com
```

What do you think about the above changes? Mainly add maintainer, source, remove comments and `apiVersion: v1` to be consistent with the other charts in the catalog

File: bitnami/contour/README.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+## TL;DR;
+
+```console
+$ helm repo add bitnami https://charts.bitnami.com/bitnami
+$ helm install my-release bitnami/contour
+```
+
+## Introduction
+
+This chart bootstraps a [Contour](https://projectcontour.io) Ingress Controller Deployment and a [Envoy Proxy](https://www.envoyproxy.io) Daemonset on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
```
Comment (carrodher) [inline bitnami/contour/README.md]:
```suggestion
## Introduction

Bitnami charts for Helm are carefully engineered, actively maintained and are the quickest and easiest way to deploy containers on a Kubernetes cluster that are ready to handle production workloads.

This chart bootstraps a [Contour](https://projectcontour.io) Ingress Controller Deployment and a [Envoy Proxy](https://www.envoyproxy.io) Daemonset on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.

Bitnami charts can be used with [Kubeapps](https://kubeapps.com/) for deployment and management of Helm Charts in clusters. This Helm chart has been tested on top of [Bitnami Kubernetes Production Runtime](https://kubeprod.io/) (BKPR). Deploy BKPR to get automated TLS certificates, logging and monitoring for your applications.
```

File: bitnami/contour/README.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+| `prometheus.serviceMonitor.enabled`                | Specify if a servicemonitor will be deployed for prometheus-operator.                                  | `true`                                                  |
+| `prometheus.serviceMonitor.jobLabel`               | Specify the jobLabel to use for the prometheus-operator                                                | `contour`                                               |
+| `prometheus.serviceMonitor.interval`               | Specify the scrape interval if not specified use defaul prometheus scrapeIntervall                     | `""`                                                    |
+| `prometheus.serviceMonitor.metricRelabelings`      | Specify additional relabeling of metrics.                                                              | `[]`                                                    |
+| `prometheus.serviceMonitor.relabelings`            | Specify general relabeling.                                                                            | `[]`                                                    |
+
+
```
Comment (carrodher) [inline bitnami/contour/README.md]: > Specify each parameter using the `--set key=value[,key=value]` argument to `helm install`. For example,
> 
> ```bash
> $ helm install my-release \
>   --set envoy.readynessProbe.successThreshold=5 \
>     bitnami/contour
> ```
> 
> The above command sets the `envoy.readynessProbe.successThreshold` to `5`.

Sorry, suggestions don't work with code blocks inside code blocks

File: bitnami/contour/README.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+## Configuration
+
```
Comment (carrodher) [inline bitnami/contour/README.md]:
```suggestion
## Configuration and installation details

### [Rolling VS Immutable tags](https://docs.bitnami.com/containers/how-to/understand-rolling-tags-containers/)

It is strongly recommended to use immutable tags in a production environment. This ensures your deployment does not change automatically if the same tag is updated with a different image.

Bitnami will release a new chart updating its containers if a new version of the main container, significant changes, or critical vulnerabilities exist.

```

File: bitnami/contour/values.yaml [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+rbac:
+  # create specifies whether to install and use RBAC rules.
+  create: true
+
+contour:
+  image:
+    registry: docker.io
+    repository: projectcontour/contour
+    tag: v1.2.1
```
Comment (carrodher) [inline bitnami/contour/values.yaml]: Once finished the container on our side, it should be something like:
```suggestion
    registry: docker.io
    repository: bitnami/contour
    tag: 1.2.1-debian-10-r0
```

File: bitnami/contour/values.yaml [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+      create: true
+      # The name of the ServiceAccount to use.  If not set and create is
+      # true, a name is generated using the fullname template
+      name: ""
+
+envoy:
+  image:
+    registry: docker.io
+    repository: envoyproxy/envoy
+    tag: v1.13.1
```
Comment (carrodher) [inline bitnami/contour/values.yaml]: Once finished the image on our side, it should be looks like:
```suggestion
    registry: docker.io
    repository: bitnami/envoy
    tag: 1.13.1-debian-10-r0
```

BTW, we planned for the next weeks the contour and metallb image, this one was not on our radar. So I will add it to our TODO list

**Comment by carrodher** [comment]
Pinging other colleagues that would be happy to take a look @juan131 @javsalgar @miguelaeh

**Comment by Cellebyte** [comment]
@carrodher  added your changes to the PR

**Comment by Cellebyte** [comment]
@carrodher I think I need to add helmV2 support.

**Comment by carrodher** [comment]
Hi, sorry, I missed this notification I don't know why.

> @carrodher I think I need to add helmV2 support.
Yes, at this point we are still supporting Helm 2 and Helm 3

**Comment by Cellebyte** [comment]
@carrodher I implemeted Helm v2 support.

**Comment by carrodher** [comment]
Hi @Cellebyte, thanks for your efforts. I have been talking with my team and we've reserved some time for reviewing the PR deeply during our next iteration (starting on Monday). So we will provide valuable feedback soon.

**Comment by Cellebyte** [comment]
@carrodher There was some stuff mentioned in the MetalLB PR which also apply to this PR I will try to fix them before the next review cycle.

**Comment by javsalgar** [comment]
Thanks @Cellebyte ! We expect the Contour container to be out very soon :D

**Comment by AntonioMeireles** [comment]
@javsalgar 

(hi!) any progress on this and/or any way the community could help speed up things ? 

Thanks in advance!

**Comment by carrodher** [comment]
Hi, sorry for the delay, we are already finishing the docker image on our side. As a new version appears in the meantime we are working on release directly the new version (`1.5.0`).

In the same way, I am going to test the chart but using the Contour Docker image (the already existing one) instead of the one we are building, so everything should work with the upstream image and then the replacement with the Bitnami one should be smooth.

**Comment by carrodher** [comment]
Hi, we already released the Bitnami Contour image:
- GitHub: https://github.com/bitnami/bitnami-docker-contour
- DockerHub: https://hub.docker.com/repository/docker/bitnami/contour
It's fully integrated into our release pipeline, just in case, you want to give it a try. Then we're also working in the envoy one that should be ready soon.

Regarding the chart itself, I was doing some tests but I found an issue with CRDs (even disabling them), did you face this issue in your deployments?
```console
$ helm install contour . --set contour.image.repository=projectcontour/contour --set contour.image.tag=latest --set envoy.image.repository=envoyproxy/envoy --set envoy.image.tag=latest --set createCustomResource=false
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"

Error: rendered manifests contain a resource that already exists. Unable to continue with install: CustomResourceDefinition "tlscertificatedelegations.projectcontour.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key "app.kubernetes.io/managed-by": must be set to "Helm"; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "contour"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "default"
```

**Comment by Cellebyte** [comment]
@carrodher I never tried installing a crd with helm 2 I think I need to add the labels template into the crd. There is also som work to do for the contour chart.

**Comment by carrodher** [comment]
Hi @Cellebyte, thanks for the latest changes. Now I am able to install the chart using the new bitnami contour image (still using the upstream envoy one):
```console
$ helm install contour . --set envoy.image.repository=envoyproxy/envoy --set envoy.image.tag=latest --set createCustomResource=false
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
NAME: contour
LAST DEPLOYED: Mon Jun  8 16:17:52 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None

$ helm list
NAME         	NAMESPACE	REVISION	UPDATED                                  	STATUS  	CHART           	APP VERSION
contour      	default  	1       	2020-06-08 16:17:52.634030742 +0000 UTC  	deployed	contour-0.3.0   	1.5.0

$ kubectl get pods
NAME                              READY   STATUS      RESTARTS   AGE
contour-contour-c4bb5f5f5-m9xvb   1/1     Running     0          6m35s
contour-contour-c4bb5f5f5-sjhvf   1/1     Running     0          6m35s
contour-contour-certgen-cq4jh     0/1     Completed   0          6m35s
```

We are working right now on the envoy image, so we are close to being able to fully include this chart in our release pipeline. In the meantime, are you thinking about adding some NOTES.txt with some basic instructions to be shown during the installation?

**Comment by Cellebyte** [comment]
@carrodher Is something missing here?

**Comment by carrodher** [comment]
Hi, not on your side. Once deployed all the images we are working on creating some tests to fully integrate the release of the chart in our CI/CD pipeline.

Manually everything seems to be working fine with the bitnami images:
```console
$ helm list
NAME      	NAMESPACE	REVISION	UPDATED                                  	STATUS  	CHART                    	APP VERSION
contour   	default  	1       	2020-06-17 16:24:48.951726459 +0000 UTC  	deployed	contour-0.3.0            	1.5.0

$ kubectl get pods
NAME                                          READY   STATUS              RESTARTS   AGE
contour-contour-7546d4df7-94p2w               1/1     Running             0          4m58s
contour-contour-7546d4df7-nrv4g               1/1     Running             0          4m58s
contour-contour-certgen-nkps7                 0/1     Completed           0          4m58s
```

Once created the tests and the integration, we can merge this PR and the chart is going to receive automatically the image updates.

**Comment by javsalgar** [comment]
Hi! I updated your branch with some fixes. However, it would get updated here

**Comment by javsalgar** [comment]
Managed to fix it, trying again.

**Comment by Cellebyte** [comment]
@javsalgar Thank you :)

**Comment by javsalgar** [comment]
Merged. Thank you very much for the contribution!!

**Comment by Cellebyte** [comment]
You are welcome.

**Pull request merged**